File: 7-janerose-projects/1-getting-started/nightlight_python/src/main.py
# Code for the server that processes telemetry from the nightlight device
# and sends commands back to it based on light levels.
import json
import time

#  this library automatically checks if there is a network connection
from paho.mqtt import client as mqtt_client

CLIENT_ID = "2fe8f0bf6f440b14087f4b66ae305e13bd5b1b5d39480a8995be1834abf3ff46"

CLIENT_TELEMETRY_TOPIC = CLIENT_ID + r"/telemetry"
SERVER_COMMAND_TOPIC = CLIENT_ID + r"/command"

CLIENT_NAME = CLIENT_ID + r"/nightlight_server"
BROKER_URL = "test.mosquitto.org"

LIGHT_THRESHOLD = 300

mqtt_client = mqtt_client.Client(callback_api_version=2, client_id=CLIENT_NAME)
mqtt_client.connect(BROKER_URL)

mqtt_client.loop_start()


def receive_telemetry(client, userdata, message):
    payload = json.loads(message.payload.decode())
    print("Received telemetry on topic:", message.topic)
    process_telemetry(payload)


def process_telemetry(payload):
    print("Processing telemetry...")
    INVERT = False
    light_val = int(payload.get('lightIn', 0))
    print("lightIn =", light_val, "threshold =", LIGHT_THRESHOLD, "invert =", INVERT)
    if INVERT:
        led_on = light_val > LIGHT_THRESHOLD
    else:
        led_on = light_val < LIGHT_THRESHOLD
    command = {'LedOn': led_on}
    mqtt_client.publish(SERVER_COMMAND_TOPIC, json.dumps(command))
    print("Sent command:", command)


mqtt_client.on_message = receive_telemetry
mqtt_client.subscribe(CLIENT_TELEMETRY_TOPIC)

while True:
    time.sleep(2)

File: 7-janerose-projects/1-getting-started/nightlight_python/src/main.cpp
// Runs on an ESP32-C3-DevKitC-02 board, reads a light sensor value, sends telemetry to an MQTT broker,
// and receives commands to control an external LED based on the telemetry data.
#include <Arduino.h>
#include <ArduinoJson.h>
#include <PubSubClient.h>
#include <WiFi.h>

#include "config.h"

WiFiClient wifiClient;
PubSubClient mqttClient(wifiClient);

const int LIGHT_SENSOR_PIN = 2; // GPIO pin for light sensor
const int LED_PIN = 13;         // external LED wired to GPIO13 (series resistor to GND)

// put function declarations here:
void connectMqttClient();
void clientCallback(char *topic, uint8_t *payload, unsigned int length);

void connectWiFi()
{
  Serial.print("Connecting to WiFi SSID: ");
  Serial.println(SSID);

  WiFi.begin(SSID, PASSWORD);

  unsigned long start = millis();
  const unsigned long timeout = 20000; // 20s timeout

  while (WiFi.status() != WL_CONNECTED && (millis() - start) < timeout)
  {
    Serial.print(".");
    delay(500);
  }

  if (WiFi.status() == WL_CONNECTED)
  {
    Serial.println();
    Serial.print("Connected! IP address: ");
    Serial.println(WiFi.localIP());
  }
  else
  {
    Serial.println();
    Serial.print("Failed to connect, status=");
    Serial.println(WiFi.status());
    // ESP32 WiFi status codes:
    // 0 = WL_IDLE_STATUS, 1 = WL_NO_SSID_AVAIL, 3 = WL_CONNECTED,
    // 4 = WL_CONNECT_FAILED, 6 = WL_DISCONNECTED
  }
}

void createMqttClient()
{
  mqttClient.setServer(BROKER_URL.c_str(), 1883);
  connectMqttClient();
  // Call the callback function whenever incoming messages are received
  mqttClient.setCallback(clientCallback);
}

void setup()
{
  Serial.begin(115200); // ESP32 typically uses 115200 baud rate
  while (!Serial)
    ;
  delay(1000);

  // Configure pins
  pinMode(LIGHT_SENSOR_PIN, INPUT);
  pinMode(LED_PIN, OUTPUT);

  // Initialize WiFi mode
  WiFi.mode(WIFI_STA);

  connectWiFi();
  createMqttClient();
}

void connectMqttClient()
{
  // recursively checks the connection to the MQTT broker and reconnects if the connection is lost
  while (!mqttClient.connected())
  {
    Serial.print("Attempting MQTT connection...");
    if (mqttClient.connect(CLIENT_NAME.c_str()))
    {
      Serial.println("connected");
      mqttClient.subscribe(SERVER_COMMAND_TOPIC.c_str());
      Serial.print("Subscribed to: ");
      Serial.println(SERVER_COMMAND_TOPIC);
    }
    else
    {
      Serial.print("failed, rc=");
      Serial.print(mqttClient.state());
      Serial.println(" try again in 5 seconds");
      delay(5000);
    }
  }
}

void clientCallback(char *topic, uint8_t *payload, unsigned int length)
{
  // Convert payload to string
  char buff[length + 1];
  for (unsigned int i = 0; i < length; i++)
    buff[i] = (char)payload[i];
  buff[length] = '\0';

  Serial.print("Message arrived: ");
  Serial.println(buff);

  DynamicJsonDocument doc(1024);
  DeserializationError error = deserializeJson(doc, buff);
  if (error)
  {
    Serial.print("JSON parse failed: ");
    Serial.println(error.c_str());
    return;
  }

  JsonObject obj = doc.as<JsonObject>();
  bool LedOn = obj["LedOn"];

  Serial.print("Command received - LedOn: ");
  Serial.println(LedOn ? "TRUE" : "FALSE");

  if (LedOn)
  {
    digitalWrite(LED_PIN, HIGH); // turn the LED on
    Serial.println("LED turned ON");
  }
  else
  {
    digitalWrite(LED_PIN, LOW); // turn the LED off
    Serial.println("LED turned OFF");
  }
}

void loop()
{
  // Maintain MQTT connection
  if (!mqttClient.connected())
  {
    connectMqttClient();
  }
  mqttClient.loop();

  // Check WiFi connection
  if (WiFi.status() != WL_CONNECTED)
  {
    Serial.println("WiFi disconnected, reconnecting...");
    connectWiFi();
  }

  // Read light sensor value
  int lightIn = analogRead(LIGHT_SENSOR_PIN);
  Serial.print("Light value (local): ");
  Serial.println(lightIn);

  // Create and send telemetry
  DynamicJsonDocument doc(1024);
  doc["lightIn"] = lightIn;
  doc["timestamp"] = millis();

  String telemetry;
  serializeJson(doc, telemetry);

  Serial.print("Sending telemetry: ");
  Serial.println(telemetry);

  mqttClient.publish(CLIENT_TELEMETRY_TOPIC.c_str(), telemetry.c_str());

  delay(2000);
}